path_planning.py

```diff
@@ -0,0 +1,321 @@
+        
+        neighbors = find_neighbors(terrain, node)       
+        for neighbor, neighbor_cost in neighbors:
+            if(neighbor in  best_costs):
+                #print("found=" , neighbor)     
+                continue
+            else:    
+                    cost,point = best_costs.get(node)#print(neighbor ,"newcost", cost + neighbor_cost)
+                    best_costs[neighbor] = ( cost + neighbor_cost ,  node )                   
+                    heapq.heappush (open_list,  ( manhattan_distance(neighbor, goal),  neighbor ) )     
+                    
+    if(found == False) :                   
+        return (visit_count, -1, None)
+    
+
+   
+    node = goal
+    bestcost,point = best_costs.get(node)   
+    retpath: List[Point] = []
+    while(1):       
+        cost,point = best_costs.get(node)          
+        retpath.insert(0,node)
+        if(compare_points(point,start)) :
+            retpath.insert(0,point)
+            break;
+        node = point
+     
+    print("bestcost" ,bestcost    )
+    print("path" ,retpath    )
+   
+ 
+    return (visit_count, bestcost, retpath)    
+
+ 
+
+def dijkstra(terrain:Map, start:Point, goal:Point) -> Tuple[int, int, Optional[List[Point]]]:
+    open_list: List[Tuple[int, Point]] = [(0, start)]         
+    best_costs: Dict[Point, Tuple[int, Point]] = {start:(0, start)}
+    visit_count = 0
+    found = False
+    while open_list:
+        # Dijkstra's search uses the priority queue data structure
+        cost, node = heapq.heappop(open_list)       
+        #print(node)
+       
+        if(compare_points(node,goal)):
+            #print ("found it") 
+            found = True;
+            break;
+                
+        visit_count += 1
+         
+        neighbors = find_neighbors(terrain, node)
+        for neighbor, neighbor_cost in neighbors:
+            if(neighbor in  best_costs):
+                #print("found=" , neighbor)     
+                ncost,npoint = best_costs.get(neighbor)
+                if(ncost > cost + neighbor_cost) :        
+                    #print("update cost for ", neighbor ,"newcost", cost + neighbor_cost)
+                    best_costs[neighbor] = ( cost + neighbor_cost ,  node )   
+                    heapq.heappush (open_list,  (cost + neighbor_cost,  (neighbor) ) )     
+            else:    
+                    #print(neighbor ,"newcost", cost + neighbor_cost)
+                    best_costs[neighbor] = ( cost + neighbor_cost ,  node )    
+                    heapq.heappush (open_list,  (cost + neighbor_cost,  (neighbor) ) )       
+                    
+    if(found == False) :                   
+        return (visit_count, -1, None)
+    
+   
+    node = goal
+    bestcost,point = best_costs.get(node)   
+    retpath: List[Point] = []
+    while(1):       
+        cost,point = best_costs.get(node)          
+        retpath.insert(0,node)
+        if(compare_points(point,start)) :

```

Comment:
Might be a little cleaner to use `not compare_points(point, start)` as the condition for the while loop and then manually insert the start at the end instead of breaking like this. I wonder if Python has a do/while, would make this nicer